### PR TITLE
Implement a few ice-ocean examples

### LINKED ISSUE #17: Implement an idealized coupled ice-ocean column model with SlabSeaIceModel
This example is an important step towards three-dimensional coupling. We'll sort out passing the latent heat from the ocean model to the ice in freezing or ice-covered areas, and passing the excess latent heat from the ice into the ocean when the ice disappears.

This task can be broken up into a few subtasks which I propose we complete in one PR:

1. thermally coupled ice-ocean simulation
2. thermally and salinity-coupled ice-ocean simulation
3. mechanically, thermally, and salinity coupled ice-ocean simulation

Once the third is finished then we have a viable coupled setup.

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements a few simple ice-ocean examples in the new directory `validation/ice_ocean_model/`. That directory also contains a prototype implementation of `IceOceanModel`, an abstraction for a coupled-ice-ocean model that contains an ocean simulation, ice simulation, and coupling parameterizations.

The prototype `IceOceanModel` works on GPUs and implements the "ice bath" parameterization for ice-ocean heat fluxes.

Closes #17

### CODE REVIEW

**Comment by glwagner** [comment]
Here's a plot from an "ocean mixed layer" case:

image

This model is relatively simple; since there are no horizontal heat fluxes in the ocean, there is no ice-ocean heat flux while the ocean is ice-covered. The ocean gets saltier when ice is forming, and that salt is returned to the ocean when the ice melts.

**Comment by glwagner** [comment]
Here's a similar example as above, but replacing the slab ocean with a column model. Convection mixes the column to the bottom during the initial cooling, so the column behaves similarly as the slab. However, during the warming phase the heat and salinity melt are trapped at the surface, so the ocean surface warms much more.

https://github.com/CliMA/ClimaSeaIce.jl/assets/15271942/69d639ce-9a1b-471c-aedf-4ff1b7c31d93

**Comment by glwagner** [comment]
The coupled cases required a substantial amount of code to implement a coupled model abstraction. I'm going to move the coupled model abstraction to `ClimaOcean`, so when we merge this PR the example will depend on ClimaOcean. Possibly we want to have a simple coupled model implementation that is independent of ClimaOcean --- perhaps a slab freshwater ocean, so we can ignore salinity fluxes.

**Comment by glwagner** [comment]
A few more examples...

# Baroclinic ice cube

https://github.com/CliMA/ClimaSeaIce.jl/assets/15271942/c0bff019-3c78-4e3c-a272-475e2356fa51

# Salty baroclinic ice cube

https://github.com/CliMA/ClimaSeaIce.jl/assets/15271942/302ad9ca-f133-446a-b6a1-c0ff73cf17a6

**Comment by navidcy** [comment]
why these examples are not part of the docs? are they computationally intense?

**Comment by glwagner** [comment]
> why these examples are not part of the docs? are they computationally intense?

Yes, like 256^3 and running on GPU.

**Comment by glwagner** [comment]
They also use `IceOceanModel` which is not part of the source code. `IceOceanModel` is prototyped here, but will be moved to `ClimaOcean`. Its' not ready for docs.

**Review by navidcy — APPROVED** [review]
these examples look sweet!

File: validation/ice_ocean_model/ice_ocean_model.jl [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+function time_step!(coupled_model::IceOceanModel, Δt; callbacks=nothing)
+    ocean = coupled_model.ocean
+    ice = coupled_model.ice
+    liquidus = ice.model.phase_transitions.liquidus
+    grid = ocean.model.grid
+    ice.Δt = Δt
+    ocean.Δt = Δt
+
+    fill_halo_regions!(h)
```
Comment (simone-silvestri) [inline validation/ice_ocean_model/ice_ocean_model.jl:161]: where is this `h` coming from?

Comment (glwagner) [reply validation/ice_ocean_model/ice_ocean_model.jl:161]: Hm, there should be an

```julia
h = ice.model.ice_thickness
```

before.

it's taking it from the global scope here so it just happened to work.

Comment (glwagner) [reply validation/ice_ocean_model/ice_ocean_model.jl:161]: fixed here https://github.com/CliMA/ClimaSeaIce.jl/pull/20

**Pull request merged**